core/scripts/liquidity-mining/test/CalculateBalancerLPProviders.js

```diff
@@ -0,0 +1,337 @@
+    beforeEach(async function() {
+      bpToken = await Token.new("BPT", "BPT", 18, {
+        from: contractCreator
+      });
+      await bpToken.addMember(1, contractCreator, {
+        from: contractCreator
+      });
+    });
+
+    it("Correctly splits payout over all liquidity providers (balanced case)", async function() {
+      // Create 10e18 tokens to distribute, sending 2e18 to each liquidity provider.
+      for (shareHolder of shareHolders) {
+        await bpToken.mint(shareHolder, toWei("2"), { from: contractCreator });
+      }
+      // Create an object to store the payouts for a given block. This should be an object with key being the
+      // shareholder address and value being their respective payout.
+      let shareHolderPayout = {};
+      for (shareHolder of shareHolders) {
+        shareHolderPayout[shareHolder] = toBN("0");
+      }
+
+      const blockNumber = await web3.eth.getBlockNumber();
+      // Distribute 5 tokens per snapshot. As there are 5 shareholders, each holding 2e18 tokens, we should expect that
+      // each shareholder should be attributed 1/5 of the total rewards, equalling 1e18 each.
+      const tokensPerSnapShot = toWei("5");
+
+      // Call the `_updatePayoutAtBlock` to get the distribution at a given `blockNumber`.
+      let bPool = new web3.eth.Contract(bpToken.abi, bpToken.address);
+      const payoutAtBlock = await _updatePayoutAtBlock(bPool, blockNumber, shareHolderPayout, tokensPerSnapShot);
+
+      // Validate that each shareholder got the right number of tokens for the given block. Expecting 1/5 of all rewards
+      // per shareholder as equal token distribution.
+      const expectedPayoutPerShareholder = toBN(tokensPerSnapShot).divn(shareHolders.length);
+      for (shareholder of shareHolders) {
+        assert.equal(payoutAtBlock[shareHolder].toString(), expectedPayoutPerShareholder.toString());
+      }
+    });
+    it("Correctly splits payout over all liquidity providers (unbalanced case)", async function() {
+      // Create 10e18 tokens to distribute, sending all to one liquidity provider.
+      await bpToken.mint(shareHolders[0], toWei("10"), { from: contractCreator });
+
+      // Create an object to store the payouts for a given block. This should be an object with key being the
+      // shareholder address and value being their respective payout.
+      let shareHolderPayout = {};
+      for (shareHolder of shareHolders) {
+        shareHolderPayout[shareHolder] = toBN("0");
+      }
+
+      const blockNumber = await web3.eth.getBlockNumber();
+      // Distribute 5 tokens per snapshot. There is only one shareholder who holds tokens at the given block number. They
+      // should exclusively receive all token payouts.
+      const tokensPerSnapShot = toWei("5");
+
+      // Call the `_updatePayoutAtBlock` to get the distribution at a given `blockNumber`.
+      let bPool = new web3.eth.Contract(bpToken.abi, bpToken.address);
```

Comment:
nit: instead of creating a new `web3.eth.Contract`, can you just use `bpToken.contract` to get the web3 instance sitting behind truffle's contract abstraction? Here and elsewhere.

Reply:
ye, that's way cleaner. I couldn't figure out the syntax for this but it turns out it's dead simple! thanks for pointing this out :)